### PR TITLE
feat(commerce-ssr): add parameter manager

### PR DESCRIPTION
**Pull request by fbeaudoincoveo**
https://coveord.atlassian.net/browse/KIT-3462

This is essentially adapted from the parameter serializer we have for non-commerce SSR.

### CODE REVIEW

File: packages/headless/src/features/commerce/parameters/parameters-serializer.ssr.test.ts [hunk 1]
```diff
@@ -0,0 +1,269 @@
+import {CommerceSearchParameters} from '../search-parameters/search-parameters-actions.js';
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/commerce/parameters/parameters-serializer.ssr.test.ts:1]: Adapted from https://github.com/coveo/ui-kit/blob/master/packages/headless/src/features/search-parameters/search-parameter-serializer.ssr.test.ts

File: packages/headless/src/features/commerce/parameters/parameters-serializer.ssr.ts [hunk 2]
```diff
@@ -0,0 +1,335 @@
+import {isNullOrUndefined} from '@coveo/bueno';
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/commerce/parameters/parameters-serializer.ssr.ts:1]: Adapted from https://github.com/coveo/ui-kit/blob/master/packages/headless/src/features/search-parameters/search-parameter-serializer.ssr.ts

Where possible and simple, I'm reusing util functions.

The most different part was dealing with sort criteria.

File: packages/samples/headless-ssr-commerce/app/products/[productId]/page.tsx [hunk 3]
```diff
@@ -11,19 +12,25 @@ export default async function ProductDescriptionPage({
...
   const navigatorContext = new NextJsNavigatorContext(headers());
-  searchEngineDefinition.setNavigatorContextProvider(() => navigatorContext);
+  standaloneEngineDefinition.setNavigatorContextProvider(
+    () => navigatorContext
+  );
 
   // Fetches the static state of the app with initial state (when applicable)
-  const staticState = await searchEngineDefinition.fetchStaticState();
+  const staticState = await standaloneEngineDefinition.fetchStaticState({
+    controllers: {parameterManager: {initialState: {parameters: {}}}},
```
Comment (fbeaudoincoveo) [inline packages/samples/headless-ssr-commerce/app/products/[productId]/page.tsx:21]: Ideally we wouldn't need to do that with the standaloneEngineDefinition... I'll see if I can fix this.

File: packages/samples/headless-ssr-commerce/components/cart.tsx [hunk 4]
```diff
@@ -46,7 +46,6 @@ export default function Cart() {
             <p>
               <span>Price: </span>
               <span>{formatCurrency(item.price, language(), currency())}</span>
-              <span>{item.price}</span>
```
Comment (fbeaudoincoveo) [inline packages/samples/headless-ssr-commerce/components/cart.tsx:49]: Unrelated, but this was a quick fix so I did the boyscout ✌️

File: packages/headless-react/src/ssr-commerce/client-utils.ts [hunk 5]
```diff
@@ -63,3 +70,27 @@ export function useSyncMemoizedStore<T>(
 
   return snapshot.current;
 }
+
+function getUrl() {
+  if (typeof window === 'undefined') {
```
Comment (louis-bompart) [inline packages/headless-react/src/ssr-commerce/client-utils.ts:75]: Do we need this despite the `use client`?

File: packages/samples/headless-ssr-commerce/components/parameter-manager.tsx [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+  const correctedUrl = useMemo(() => {
+    if (!historyRouter.url) {
+      return null;
+    }
+    const newURL = new URL(historyRouter.url);
+    const {serialize} = buildSSRCommerceSearchParameterSerializer();
+
+    return serialize(state.parameters, newURL);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (louis-bompart) [inline packages/samples/headless-ssr-commerce/components/parameter-manager.tsx:33]: Which deps are we missing, and why :suspect: ?

File: packages/samples/headless-ssr-commerce/components/parameter-manager.tsx [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    const {pathname} = new URL(correctedUrl);
+    if (pathname !== document.location.pathname) {
+      return;
+    }
+    const isStaticState = controller === undefined;
+
+    historyRouter[isStaticState ? 'replace' : 'push'](correctedUrl);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (louis-bompart) [inline packages/samples/headless-ssr-commerce/components/parameter-manager.tsx:48]: same

**Review by louis-bompart — COMMENTED** [review]
Couple ❓about the non exhaustive deps, it's often a big smell in React.

File: packages/samples/headless-ssr-commerce/components/parameter-manager.tsx (hunk 6 above)
Comment (fbeaudoincoveo) [reply packages/samples/headless-ssr-commerce/components/parameter-manager.tsx:33]: This is the fruit of copy-pasting

I added the missing deps and it's all working fine.

File: packages/samples/headless-ssr-commerce/components/parameter-manager.tsx (hunk 7 above)
Comment (fbeaudoincoveo) [reply packages/samples/headless-ssr-commerce/components/parameter-manager.tsx:48]: same response :D

File: packages/samples/headless-ssr-commerce/components/parameter-manager.tsx (hunk 6 above)
Comment (fbeaudoincoveo) [reply packages/samples/headless-ssr-commerce/components/parameter-manager.tsx:33]: Upon further investigation, there's a reason for this: including the deps prevents navigation with the browser's back/forward buttons. I'm not entirely sure why. @y-lakhdar maybe you have some insight into this?

Comment (fbeaudoincoveo) [reply packages/samples/headless-ssr-commerce/components/parameter-manager.tsx:33]: For this one, `useMemo` recomputes the memoized value when one of the deps has changed, so I think that would explain it.

**Comment by fbeaudoincoveo** [comment]
Replaced with:

https://github.com/coveo/ui-kit/pull/4759
https://github.com/coveo/ui-kit/pull/4770

**Pull request closed without merging**